linker/shared/src/main/scala/org/scalajs/linker/frontend/modulesplitter/MinModuleSplitter.scala

```diff
@@ -0,0 +1,206 @@
+
+    new MinModuleAnalyzer(unit, symbolRequirements).analyze(moduleSpecs)
+  }
+}
+
+private object MinModuleAnalyzer {
+  private final class Node(val clazz: ClassName, val index: Int, var compId: Int)
+  private sealed trait Result
+  private final case class Partial(lowlink: Int, nodes: Vector[Node], deps: Set[Int]) extends Result
+  private final case class Component(id: Int) extends Result
+}
+
+private class MinModuleAnalyzer(
+    unit: LinkingUnit, symbolRequirements: SymbolRequirement) {
+  import MinModuleAnalyzer._
+
+  private val dependencyAnalyzer = new DependencyAnalyzer(unit, symbolRequirements)
+  import dependencyAnalyzer._
+
+  private[this] var nextIndex = 0
+  private[this] val visitedNodes = mutable.Map.empty[ClassName, Node]
+  private[this] val internalModules = mutable.Map.empty[Int, ModuleSet.Module]
+
+  def analyze(moduleSpecs: List[ModuleSpec]): ModuleSet = {
+    def hasAnyDefinitions(c: LinkedClass) = {
+      c.fields.nonEmpty ||
+      c.methods.nonEmpty ||
+      c.exportedMembers.nonEmpty ||
+      c.hasInstanceTests ||
+      c.hasRuntimeTypeInfo
+    }
+
+    unit.classDefs.iterator
+      .filter(hasAnyDefinitions(_))
+      .map(_.className)
+      .filter(!visitedNodes.contains(_))
+      .foreach(strongconnect(_))
+
+    val abstractClasses = unit.classDefs
+      .filter(c => !visitedNodes.contains(c.className))
+
+    val badAbstractClasses = abstractClasses.filter(hasAnyDefinitions(_))
+    assert(badAbstractClasses.isEmpty, {
+      val names = badAbstractClasses.map(_.fullName).mkString(", ")
+      "Found abstract classes that contain code: $names"
+    })
+
+    val modules = (
+        leafModules(moduleSpecs, unit) ++
+        internalModules.values.toList
+    )
+
+    new ModuleSet(unit.coreSpec, modules, abstractClasses, unit.isParentDataAccessed)
+  }
+  
+  private def leafModules(moduleSpecs: List[ModuleSpec],
+      unit: LinkingUnit): List[ModuleSet.Module] = {
+    val analyzer = new SelectorAnalyzer(moduleSpecs)
+    val exportsPerModule = analyzer.classifyTopLevelExports(unit.topLevelExports)
+    val staticInitializersPerModule = analyzer.classifyStaticInitializers(
+        unit.classDefs.filter(_.hasStaticInitializer).map(_.className))
+
+    for {
+      moduleSpec <- moduleSpecs
+    } yield {
+      val topLevelExports = exportsPerModule.getOrElse(moduleSpec.id, Nil)
+      val staticInitializers = staticInitializersPerModule.getOrElse(moduleSpec.id, Nil)
+
+      val internalDeps = {
+        val staticDeps =
+          topLevelExports.flatMap(staticDependencies(_)) ++
+          moduleSpec.initializers.flatMap(staticDependencies(_)) ++
+          staticInitializers ++
+          Iterable(ObjectClass)
+
+        staticDeps
+          .map(className => internalModules(visitedNodes(className).compId).id)
+          .toSet
+      }
+
+      val externalDeps =
+        topLevelExports.flatMap(externalDependencies(_)).distinct.toList
+
+      // TODO: Filter empty modules.
```

Comment:
Unsure what to do here. Should an empty module cause a failure? Should it cause an empty file? Should it cause no file?

Reply:
It seems to me that the most reasonable outcome would be an empty file. If a module has been requested by the user, and it ends up being empty, why should we get rid of it or consider that an error?

Reply:
> why should we get rid of it or consider that an error?

It might hint at a configuration error. But yes, probably an empty module is OK.